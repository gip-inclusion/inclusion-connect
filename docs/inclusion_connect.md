# Utiliser Inclusion Connect pour authentifier des utilisateurs

Afin d'implémenter Inclusion Connect, il faut un compte pour votre <abbr>FS</abbr> (Fournisseur de Service) sur les plateformes d'intégration et de production.

- Vous pouvez observer l'ensemble des parcours utilisateurs à [cette page](user_journey.md)
- Pour plus de détail sur les différents _endpoints_ vous pouvez continuer à lire cette page.

Contactez l'équipe du projet pour obtenir les variables de production et de recette.

## Nos endpoints

Le format des urls est le suivant :

|               |                                       |
|           --- | ---                                   |
| Authorization | https://{hostname}/auth/authorize     |
| Registration  | https://{hostname}/auth/register      |
| Activation    | https://{hostname}/auth/activate      |
| Token         | https://{hostname}/auth/token         |
| UserInfo      | https://{hostname}/auth/userinfo      |
| Logout        | https://{hostname}/auth/logout        |


D'un point de vue OpenID Connect, les end-points Authorization, Registration et Activation
sont identiques. La seule chose qui change est si l'utilisateur n'est pas connecté, dans ce cas on
le redirige vers une vue différente pour chaque end-point.

Le hostname dépend de l'instance utilisée :
- https://connect.inclusion.beta.gouv.fr/ pour la production
- https://recette.connect.inclusion.beta.gouv.fr/ pour la recette
- http://127.0.0.1:8080 pour l'environnement local

## Détail du fonctionnement

```mermaid
sequenceDiagram
Note right of Utilisateur: L'utilisateur clique sur le bouton "Inclusion Connect"
Utilisateur ->> FS: ;
FS -->> Utilisateur: Redirect 302 <IC_URL>/auth/authorize
Utilisateur ->> Inclusion Connect: GET <IC_URL>/auth/authorize.
Inclusion Connect -->> Utilisateur: Redirect 302 <FS_URL>/<URL_CALLBACK>
Utilisateur ->> FS: GET <FS_URL>/<URL_CALLBACK>
FS ->> Inclusion Connect: POST <IC_URL>/auth/token
Inclusion Connect -->> FS: HTTP Response 200
FS ->> Inclusion Connect: POST <IC_URL>/auth/userinfo
Inclusion Connect -->> FS: HTTP Response 200
FS -->> Utilisateur: Redirect 302 <FS_URL>/page_authentifée
Note right of Utilisateur: Plus tard l'utilisateur se déconnecte du fournisseur de service
Utilisateur ->> FS: ;
FS -->> Utilisateur: Redirect 302 <IC_URL>/auth/logout
Utilisateur ->> Inclusion Connect: GET <IC_URL>/auth/logout
Inclusion Connect -->> Utilisateur: Redirect 302 <FS_URL>/<POST_LOGOUT_REDIRECT_URI>
Utilisateur ->> FS: GET <FS_URL>/<POST_LOGOUT_REDIRECT_URI>
```

## Détail des flux

Le détail des flux peut être trouvé en anglais ici :
- https://openid.net/specs/openid-connect-core-1_0.html
- https://openid.net/specs/openid-connect-rpinitiated-1_0.html

### 1) Authentification / Création de compte / Activation de compte

#### Description

- Contexte : Le FS redirige l'utilisateur vers l'endpoint _Authorization_ pour engager la cinématique d'authentification.
- Origine : FS
- Cible : Inclusion Connect
- Type d'appel : redirection navigateur

#### Requête (Authentification)

- URL : `https://{hostname}/auth/authorize?response_type=code&client_id=<CLIENT_ID>&redirect_uri=<FS_URL>%2F<URL_CALLBACK>&scope=<SCOPES>&state=><STATE>&nonce=<NONCE>`
- Méthode : GET

Les paramètres sont les suivants :
- **response_type** : Le type de cinématique d'authentification utilisée. Comme nous n'utilisons que le _Authorization Code Flow_ la valeur à donner est `code`.
- **client_id** : l'identifiant du FS.
- **redirect_uri** : l'uri de redirection à laquelle sera envoyée la réponse d'Inclusion Connect.
- **scope** : les ressources auxquelles on souhaite avoir accès. Il faut utiliser `openid profile email`.
- **state** : valeur générée aléatoirement par le FS qu'Inclusion Connect renvoie tel quel dans la réponse à cet appel pour être ensuite vérifié par le FS. Il est utilisé afin d’empêcher l’exploitation de failles CSRF.
- **nonce** : valeur générée aléatoirement par le FS qu'Inclusion Connect renvoie tel quel dans la réponse à l'appel à /token, pour être ensuite vérifié par le FS. Il est utilisé pour empêcher les attaques par rejeu.
- **login_hint**: paramètre optionnel servant à pré-remplir une adresse e-mail dans les différents formulaires de connexion / création de compte. Lorsque ce paramètre est fourni, il n'est pas possible de changer l'adresse email pré-remplie.

/!\ Attention : si l'utilisateur est déjà connecté avec un autre email, Inclusion Connect acceptera automatiquement la demande de connexion, et l'email ne correspondra donc pas a celui donné dans login_hint. Il faut donc que le FS vérifie que l'email est identique de son côté.

Une fois sur Inclusion Connect, il y a 3 possibilités.
- Si c'est un nouvel utilisateur, il devra se créer un compte, puis recevra un email de vérification d'adresse email qui le renverra sur Inclusion Connect avant d'être redirigé sur **redirect_uri**.
- Si c'est un utilisateur existant non connecté, il devra entrer ses identifiants et sera ensuite redirigé sur **redirect_uri**.
- Si c'est un utilisateur déjà connecté, il sera directement redirigé sur **redirect_uri**.

#### Requête (Création de compte)

Il est possible d'utiliser l'endpoint _Registration_ pour que l'utilisateur arrive directement sur cette seconde page.

- URL : `https://{hostname}/auth/register?response_type=code&client_id=<CLIENT_ID>&redirect_uri=<FS_URL>%2F<URL_CALLBACK>&scope=<SCOPES>&state=><STATE>&nonce=<NONCE>`
- Méthode : GET

Les paramètres sont les mêmes que pour l'Authentification.

#### Requête (Activation de compte)

Il est possible d'utiliser l'endpoint _Activation_ pour que l'utilisateur arrive directement sur une page de création de compte pré-remplie.

- URL : `https://{hostname}/auth/activate?response_type=code&client_id=<CLIENT_ID>&redirect_uri=<FS_URL>%2F<URL_CALLBACK>&scope=<SCOPES>&state=><STATE>&nonce=<NONCE>`
- Méthode : GET

Les paramètres sont les mêmes que pour l'Authentification, plus :
- **login_hint**: devient obligatoire
- **firstname**: permet de pré-remplir le prénom de l'utilisateur
- **lastname**: permet de pré-remplir le nom de famille de l'utilisateur
Sur cette page, ces informations ne sont pas modifiables. Cela permet de s'assurer qu'un utilisateur existant d'une plateforme qui migre son compte à Inclusion Connect ne change pas ses informations personnelles.

#### Réponse

HTTP 302 - Redirection vers **redirect_uri***

### 2) Réponse à l'authentification

#### Description

- Contexte : L'utilisateur a complété la cinématique sur Inclusion Connect qui appelle le callback du FS avec un code d'autorisation dans l'URL.
- Origine : Inclusion Connect
- Cible : FS
- Type d'appel : redirection navigateur

#### Requête

- URL : `<FS_URL>/<URL_CALLBACK>?code=<AUTHZ_CODE>"&state=<STATE>`
- Méthode : GET

Les paramètres sont les suivants :
- **code** : code à réutiliser lors de l'appel sur le endpoint _Token_.
- **state** : la valeur passée lors de l'appel à l'endpoint _Authorisation_ et qu'il faut donc vérifier par sécurité.

### 3) Récupération d'un _Access Token_

#### Description

- Contexte : Le FS a reçu le code d'autorisation. Il doit maintenant demander un _access token_ à Inclusion Connect.
- Origine : FS
- Cible : Inclusion Connect
- Type d'appel : appel de web service

#### Requête

- URL : `https://{hostname}/auth/token`
- Méthode : POST
- Header attendu: `Content-Type: application/x-www-form-urlencoded`
- Body: au format `key1=value1&key2=value2` (adapté au content-type `application/x-www-form-urlencoded` )

Voici le payload à envoyer :
- **grant_type**' : `authorization_code` (valeur imposée par le protocol).
- **redirect_uri**' : à nouveau LA MÊME uri de redirection passée à l'endpoint _Authorization_ '<FS_URL>/<CALLBACK_URL_DATA>'.
- **client_id** : l'identifiant du FS.
- **client_secret** : le secret du FS correspondant à son identifiant.
- **code** : le code reçu en réponse à la requête d'authentification.


#### Réponse

Corps HTTP:
```
{
  'access_token': <ACCESS_TOKEN>,
  'token_type': 'Bearer',
  'expires_in': 60,
  'id_token': <ID_TOKEN>
}
```

L'_access token_ sera ensuite utilisé pour accéder à l'endpoint _UserInfo_.

L'_id token_ est un objet JWT signé qui contient notamment :
- **nonce** : la valeur transmise lors de la requête initiale qu'il faut vérifier.
- **sub** : l'identifiant unique de l'utilisateur que le FS doit conserver au cas où l'utilisateur change son adresse e-mail un jour.
- **given_name** : le prénom de l'utilisateur.
- **family_name** : son nom de famille.
- **email** : son adresse e-mail.


La signature du token est chiffrée avec l'algorithme `RS256` et il est possible de récupérer la clé publique pour vérifier la signature.

Cependant, dans le cas où l'_access token_ est  récupéré via un appel direct du _backend_ au _OpendID Connect provider_ (Inclusion Connect)
et non par exemple récupéré par le _frontend_ puis transmis au _backend_, la validation du token ne protège pas grand-chose.
En effet, si Inclusion Connect a été compromis, ou si ses réponses sont modifiées par un attaquant, alors celui-ci pourra vraisemblablement
aussi modifier la réponse de l'endpoint pour récupérer la clé publique et y mettre la sienne.

Ce cas précis est d'ailleurs mentionné dans le document de l'IETF "OAuth 2.0 for Browser-Based Apps" :
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-9.8.3

> Performing OpenID Connect using the Authorization Code flow provides the benefit of the client not needing to verify the JWT signature,
> as the ID token will have been fetched over an HTTPS connection directly from the authorization server.

### 4) Récupération des informations de l'utilisateur

Note: Cette étape est actuellement facultative car toutes les informations que l'on peut récupérer sur cet endpoint
sont déjà présentent dans l'id_token.

#### Description

- Contexte : Le FS interroge Inclusion Connect pour récupérer les informations de l'utilisateur.
- Origine : FS
- Cible : Inclusion Connect
- Type d'appel : appel de web service

#### Requête

- URL :  `https://{hostname}/auth/userinfo`
- Méthode : GET. La transmission de l'_access_token_ DOIT se faire dans le header `Authorization : Bearer <Token>`.

#### Réponse

Corps HTTP:
```
{
    'sub': <l'identifiant unique de l'utilisateur>
    'given_name': <prénom>,
    'family_name': <nom de famille>
    'email': <l'email>
}
```

Note : il faut que le FS conserve la valeur de sub en base de donnée car c'est celle-ci qui permettra d'identifier l'utilisateur s'il choisit de changer
son adresse email dans Inclusion Connect.

### 5) Déconnexion

Lorsqu'un utilisateur se déconnecte d'un FS, on s'attend à ce que le FS appelle l'endpoint _Logout_ d'Inclusion Connect.

Cela peut être fait de deux manières :

#### Sans confirmation

On redirige l'utilisateur sur `https://{hostname}/auth/logout?state=<STATE>&id_token_hint=<ID_TOKEN>&post_logout_redirect_uri=<FS_URL>%2F<POST_LOGOUT_REDIRECT_URI>`
qui sera ensuite redirigé vers l'url passée avec le paramètre **post_logout_redirect_uri**.


#### Avec confirmation

Si le `STATE` et/ou l'`ID_TOKEN` ne sont pas disponibles, il est possible de déconnecter l'utilisateur avec une redirection vers `https://{hostname}/auth/logout?client_id=<CLIENT_ID>&post_logout_redirect_uri=<FS_URL>%2F<POST_LOGOUT_REDIRECT_URI>`.
Dans ce cas, l'utilisateur devra confirmer sa volonté de se deconnecter d'Inclusion Connect et sera ensuite redirigé vers l'url passée avec le paramètre **post_logout_redirect_uri**.
