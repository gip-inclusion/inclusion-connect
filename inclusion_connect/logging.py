import logging.handlers
import threading

from elasticsearch import Elasticsearch
from elasticsearch.helpers import bulk
from pythonjsonlogger import jsonlogger

from inclusion_connect.utils.oidc import oidc_params


def log_data(request, next_url=None):
    log_data = {"ip_address": request.META["REMOTE_ADDR"]}
    params = oidc_params(request, next_url)
    try:
        log_data["application"] = params["client_id"]
    except KeyError:
        pass
    return log_data


class JsonFormatter(jsonlogger.JsonFormatter):
    def parse(self):
        # Remove the empty key "message".
        return []

    def add_fields(self, log_record, record, message_dict):
        super().add_fields(log_record, record, message_dict)
        for record_attr in ["name", "levelname"]:
            log_record[record_attr] = getattr(record, record_attr)


class ElasticSearchHandler(logging.handlers.BufferingHandler):
    timer = None

    def __init__(  # noqa: PLR0913 Too many arguments to function call.
        self,
        *,
        capacity,
        send_after_inactive_for_secs=2.0,
        host,
        index_name,
        es_timeout_secs=5,
    ):
        """
        Inspired from https://github.com/cmanaha/python-elasticsearch-logger/blob/master/cmreslogging/handlers.py.

        :param int capacity: Max number of log records before sending to ElasticSearch.
        :param float send_after_inactive_for_secs: Send to ElasticSearch if no records
            are emitted in send_after_inactive_for_secs seconds.
        """
        super().__init__(capacity)
        self.send_after_inactive_for_secs = send_after_inactive_for_secs
        self.es_client = Elasticsearch(host, http_compress=True, request_timeout=es_timeout_secs, max_retries=10)
        self.index_name = index_name

    def emit(self, record):
        if self.timer and self.timer.is_alive():
            self.timer.cancel()
        formatted_record = self.format(record)
        super().emit(formatted_record)
        self.timer = threading.Timer(self.send_after_inactive_for_secs, self.flush)
        self.timer.start()

    def send_to_elastic(self, log_buffer):
        actions = ({"_source": log} for log in log_buffer)
        bulk(client=self.es_client, actions=actions, index=self.index_name, stats_only=True)

    def flush(self):
        with self.lock:
            log_buffer = self.buffer
            self.buffer = []
        if log_buffer:
            t = threading.Thread(target=self.send_to_elastic, args=(log_buffer,), daemon=False)
            t.start()
