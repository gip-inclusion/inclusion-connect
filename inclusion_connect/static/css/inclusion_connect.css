body {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
}

main {
  flex-grow: 1;
  background-color: #FFFFFF;
  background-image: url('../img/illustration-bg-ic.svg');
  background-repeat: no-repeat;
  background-position: top -120px left;
}

.form-group > label,
.input-group > label {
  font-weight: 500;
}

.pe-connect {
  display: flex;
  justify-content: center;
  align-items: center;
}

