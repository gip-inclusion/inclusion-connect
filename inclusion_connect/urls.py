"""inclusion_connect URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/4.1/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path


urlpatterns = [
    # Admin
    path("admin/", admin.site.urls),
    # Login and registration urls
    re_path(r"^accounts/", include("inclusion_connect.accounts.urls")),
    re_path(r"^accounts/", include("django.contrib.auth.urls")),
    # OIDC urls
    re_path(r"^auth/", include("inclusion_connect.oidc_overrides.urls")),
    re_path(r"^auth/", include("oauth2_provider.urls", namespace="oauth2_provider")),
]

for realm in settings.KEYCLOAK_REALMS:
    urlpatterns.append(
        re_path(
            rf"^realms/{realm}/protocol/openid-connect/",
            include("inclusion_connect.keycloak_compat.urls", namespace=f"keycloak_compat_{realm}"),
        )
    )

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
