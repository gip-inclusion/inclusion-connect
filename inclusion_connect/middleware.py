from django.core.exceptions import PermissionDenied
from django.urls import reverse
from django.utils.cache import add_never_cache_headers
from django.utils.html import format_html


def never_cache(get_response):
    def middleware(request):
        response = get_response(request)
        if request.user.is_authenticated:
            add_never_cache_headers(response)
        return response

    return middleware


def limit_staff_users_to_admin(get_response):
    def middleware(request):
        user = request.user

        if user.is_staff and not request.path.startswith("/admin/"):
            exception = format_html(
                "Les comptes administrateurs n'ont pas accès à cette page.<br>"
                '<a href="{}">Vous pouvez-vous déconnecter ici.</a>',
                reverse("admin:logout"),
            )
            raise PermissionDenied(exception)

        return get_response(request)

    return middleware
