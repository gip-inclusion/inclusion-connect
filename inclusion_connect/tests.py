# Functional tests for all documented customer processes
import re

import pytest
from django.contrib import messages
from django.contrib.auth import get_user
from django.core import mail
from django.urls import reverse
from pytest_django.asserts import assertContains, assertRedirects

from inclusion_connect.oidc_overrides.factories import ApplicationFactory
from inclusion_connect.test import OIDC_PARAMS, oidc_flow_followup, token_are_revoked
from inclusion_connect.users.factories import DEFAULT_PASSWORD, UserFactory
from inclusion_connect.users.models import User
from inclusion_connect.utils.urls import add_url_params, get_url_params


def test_registration_endpoint(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    user = UserFactory.build()

    auth_url = reverse("oidc_overrides:registrations")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:registration"))

    response = client.post(
        response.url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
            "terms_accepted": "on",
        },
    )
    assertRedirects(response, auth_complete_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated is True
    user = User.objects.get(email=user.email)
    assert user.linked_applications.count() == 0

    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(OIDC_PARAMS["redirect_uri"])
    auth_response_params = get_url_params(response.url)
    assert user.linked_applications.count() == 1

    oidc_flow_followup(client, auth_response_params, user)


def test_activation_endpoint(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    user = UserFactory.build()

    auth_url = reverse("oidc_overrides:activation")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url, follow=True)
    assert response.status_code == 400

    auth_url = reverse("oidc_overrides:activation")
    auth_params = OIDC_PARAMS | {"email": "email", "firstname": "firstname", "lastname": "lastname"}
    auth_complete_url = add_url_params(auth_url, auth_params)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:activation"))

    response = client.post(
        response.url,
        data={
            "email": user.email,
            "first_name": user.first_name,
            "last_name": user.last_name,
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
            "terms_accepted": "on",
        },
    )
    assertRedirects(response, auth_complete_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated is True
    user = User.objects.get(email=user.email)
    assert user.linked_applications.count() == 0

    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(OIDC_PARAMS["redirect_uri"])
    auth_response_params = get_url_params(response.url)
    assert user.linked_applications.count() == 1

    oidc_flow_followup(client, auth_response_params, user)


def test_login_endpoint(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    user = UserFactory()

    auth_url = reverse("oidc_overrides:authorize")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))

    response = client.post(
        response.url,
        data={
            "email": user.email,
            "password": DEFAULT_PASSWORD,
        },
    )
    assertRedirects(response, auth_complete_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated is True
    user = User.objects.get(email=user.email)
    assert user.linked_applications.count() == 0

    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(OIDC_PARAMS["redirect_uri"])
    auth_response_params = get_url_params(response.url)
    assert user.linked_applications.count() == 1

    oidc_flow_followup(client, auth_response_params, user)


def test_login_after_password_reset(client):
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])
    user = UserFactory()

    auth_url = reverse("oidc_overrides:authorize")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))

    response = client.get(response.url)
    assertContains(response, reverse("accounts:password_reset"))

    response = client.post(reverse("accounts:password_reset"), data={"email": user.email})
    assertRedirects(response, reverse("accounts:login"))

    assert list(messages.get_messages(response.wsgi_request)) == [
        messages.storage.base.Message(
            messages.SUCCESS,
            "Si un compte existe avec cette adresse e-mail, "
            "vous recevrez un e-mail contenant des instructions pour r√©initialiser votre mot de passe.",
        ),
    ]

    reset_url_regex = reverse("accounts:password_reset_confirm", args=("string", "string")).replace("string", "[^/]*")
    reset_url = re.search(reset_url_regex, mail.outbox[0].body)[0]
    response = client.get(reset_url)  # retrieve the modified url
    response = client.post(response.url, data={"new_password1": "password", "new_password2": "password"})
    assertRedirects(response, auth_complete_url, fetch_redirect_response=False)
    assert get_user(client).is_authenticated is True

    response = client.get(auth_complete_url)
    assert response.status_code == 302
    assert response.url.startswith(OIDC_PARAMS["redirect_uri"])
    auth_response_params = get_url_params(response.url)

    oidc_flow_followup(client, auth_response_params, user)


@pytest.mark.parametrize("method", ["get", "post"])
def test_logout_no_confirmation(client, method):
    """Logout without confirmation requires the id_token"""

    user = UserFactory()
    ApplicationFactory(client_id=OIDC_PARAMS["client_id"])

    auth_url = reverse("oidc_overrides:authorize")
    auth_complete_url = add_url_params(auth_url, OIDC_PARAMS)
    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))

    response = client.post(response.url, data={"email": user.email, "password": DEFAULT_PASSWORD})
    assert get_user(client).is_authenticated is True
    response = client.get(response.url)
    auth_response_params = get_url_params(response.url)
    id_token = oidc_flow_followup(client, auth_response_params, user)

    assert get_user(client).is_authenticated is True
    logout_params = {"id_token_hint": id_token}
    logout_method = getattr(client, method)
    response = logout_method(add_url_params(reverse("oidc_overrides:logout"), logout_params))
    assert not get_user(client).is_authenticated
    assert token_are_revoked(user)

    response = client.get(auth_complete_url)
    assertRedirects(response, reverse("accounts:login"))


def test_logout_with_confirmation(client):
    # FIXME: currently not working
    pass
