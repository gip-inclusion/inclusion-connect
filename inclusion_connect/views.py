from django.shortcuts import render
from django.urls import reverse

from inclusion_connect.utils.urls import add_url_params


def csrf_failure(request, template_name="403_csrf.html", **kwargs):
    context = {
        "edit_user_info_url": add_url_params(
            reverse("accounts:edit_user_info"), {"redirect_uri": request.GET.get("referrer_uri")}
        )
    }
    return render(request, template_name, context=context, status=403)


def home(request, template_name="homepage.html", **kwargs):
    return render(request, template_name)
