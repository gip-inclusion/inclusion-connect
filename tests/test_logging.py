import logging
from unittest import mock

from freezegun import freeze_time

from inclusion_connect.logging import ElasticSearchHandler


@freeze_time("2023-05-05 11:11:11")
def test_log_formatting(snapshot):
    logger = logging.getLogger("inclusion_connect")  # Root logger for IC.
    [handler] = logger.handlers
    logger.info({"key": "value", "other_key": "value"})
    stream = handler.stream
    stream.seek(0)
    assert stream.read() == snapshot(name="log serialized as JSON with metadata")


@mock.patch("inclusion_connect.logging.bulk")
class TestElasticSearchHandler:
    def test_elastic_search_handler_sends_at_capacity(self, bulk_mock):
        handler = ElasticSearchHandler(capacity=1, index_name="test", host="https://localhost:9200")
        handler.handle(logging.LogRecord("test_logger", logging.INFO, "pathname", 1, "msg", (), None))
        bulk_mock.assert_called_once()
        handler.timer.cancel()

    def test_elastic_search_handler_after_some_time(self, bulk_mock):
        handler = ElasticSearchHandler(
            capacity=100,
            index_name="test",
            host="https://localhost:9200",
            send_after_inactive_for_secs=0,
        )
        handler.handle(logging.LogRecord("test_logger", logging.INFO, "pathname", 1, "msg", (), None))

        # This might be a bit fragile: the timer thread may execute before the next assertion.
        # Thatâ€™s unlikely, considering the amount of work the timer thread has compared to the test thread,
        # and thousands runs do not show flakiness.
        # Keeping the assertion makes the test easier to follow and more robust.
        bulk_mock.assert_not_called()

        timer_internal_event = handler.timer.finished
        # Wait for the timer thread to complete, and verify wait did not timeout.
        assert timer_internal_event.wait(1) is True
        bulk_mock.assert_called_once()
