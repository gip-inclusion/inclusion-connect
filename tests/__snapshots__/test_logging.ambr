# serializer version: 1
# name: test_log_formatting[log serialized as JSON with metadata]
  '''
  {"key": "value", "other_key": "value", "@timestamp": "2023-05-05T11:11:11Z", "name": "inclusion_connect", "levelname": "INFO"}
  INFO:inclusion_connect:{'key': 'value', 'other_key': 'value'}
  
  '''
# ---
